### ISSUE TITLE
Error "Could not acquire name on session bus"

### ISSUE THREAD

**Issue by rdiez** [neutrinolabs/xrdp#1559]
Hallo all:

I am testing a virtual machine with the new Ubuntu MATE 20.04. This system comes with xrdp 0.9.12.

I have always been jealous of Windows systems: enable an option with the mouse, and remote desktop works straight away with very few issues. On Linux, I have always have problems in this respect.

If I am logged on in the VM, then Remmina or rdesktop show error "could not acquire name on session bus".  You find many hits on the Internet about this, but no clear, final solution.

If I log off in the VM, then remote desktop works. But then, if I close the connection, I cannot log on locally anymore. With Ubuntu MATE 20.04, you get first the same error message I mentioned above, and then the desktop appears. But if you click on anything, menu items etc. hang. The desktop does not work well.

Even if you have no solution, or it is the Linux distribution's fault, I think you should mention it in the FAQ. It is a popular problem after all.

Thanks in advance,
  rdiez

**Comment by matt335672** [comment]
Well, this is sort of documented (recently) on the wiki [here](https://github.com/neutrinolabs/xrdp/wiki/Debian-dbus-user-session-package) which seems to solve the problem for 18.04

However, this isn't completely solving the problem for me on 20.04 release version. When I'm logged in to XRDP I can't log in to the console. I can however log in to the console and then log in to XRDP. 

I've not found a workaround for that one yet.

**Comment by matt335672** [comment]
Here's some more information regarding not being able to log in to the console when an XRDP session is active, and with `dbus-user-session` removed.

This is applicable to 20.04 LTS running `3.34.1-1ubuntu1` of gdm3.

If I log an XRDP session in, `loginctl` displays the following for the XRDP session:-

```bash
$ sudo loginctl
SESSION  UID USER SEAT  TTY
     10 1000 mjb        pts/0
     c5  125 gdm  seat0 tty1
     c7 1000 mjb
$ sudo loginctl show-session c7
Id=c7
User=1000
Name=mjb
Timestamp=Tue 2020-04-28 12:46:40 BST
TimestampMonotonic=3423262249
VTNr=0
Display=:11
Remote=no
Service=xrdp-sesman
Scope=session-c7.scope
Leader=12137
Audit=4294967295
Type=x11
Class=user
Active=yes
State=active
IdleHint=no
IdleSinceHint=0
IdleSinceHintMonotonic=0
LockedHint=no
```

Comparing a gdm debug log for a failed console login (XRDP logged in) and a succeeding login (XRDP not logged in), reveals the following errors in the failed login:-

```
xrdp-test-focal gdm3[2710]: GdmManager: Error while retrieving tty for session: Error getting tty for session id c7 from systemd: No data available
xrdp-test-focal gdm-x-session[6467]: Could not register display: GDBus.Error:org.freedesktop.DBus.Error.AccessDenied: No display available
```

The corresponding line in the successful log is:-
```
xrdp-test-focal gdm3[2710]: GdmSession: Setting display device: tty2
```

So it seems possible that the presence of an XRDP session without a tty is confusing gdm, and it's failing to find the local session as a result.

At this point I have to say that I don't know enough about systemd sessions to know whether the error is with sesman or with gdm3.

In summary, the current situation with 20.04 is that if an XRDP session is logged on, it can prevent console logins. This is with the `dbus-user-session` package removed as recommended.

**Comment by rdiez** [comment]
I am not sure what you mean with "console logins". Do you mean text-mode only?

I was testing with Ubuntu 20.04. If I close the remote desktop window (Remmina or rdesktop), does the remote desktop session end? Or should I expect the session to remain, in case I reconnect later? Is there a way to choose what behaviour I want?

What I saw is that, after closing the remote desktop window, local logons (using SPICE to connect to the VM virtual graphics card) no longer work correctly. You can log on, and then you get the error message I mentioned, and the desktop remains. But the MATE desktop no longer works correctly (it hangs if you click around).

**Comment by matt335672** [comment]
Sorry - by 'console logins' I mean graphical logins in the machine console. I hope that's clearer.

If you close the RDP client (whatever it is), the session should continue and you should be able to reconnect to it, potentially from another client in another location. This is how the RDP protocol is designed to work and I don't believe there is a way to make the session end automatically if you disconnect.

Have you removed the `dbus-user-session` package as suggested on the Wiki? You will need to do this to get anything working with XRDP at all. Incidentally I've just updated the Wiki with some instructions for this, along with a pointer to this issue.

**Comment by rdiez** [comment]
You are right and I am sorry I forgot to mention it: I did not replace dbus-user-session with dbus-x11 like the wiki says. I am actually afraid of doing so. In my experience, Ubuntu and its derivatives have enough bugs with the standard packages everybody use by default. I have tried changing login manager settings and packages in the past, and it was not a nice experience.

I have another question. I guess that console (text-only) sessions over SSH are not affected by these problems. I could connect over SSH and terminate or kill any RDP sessions associated to my username that may be lingering on. Or I could change to a text-mode terminal with Ctrl+Alt+F1 if I am in front of the PC in question. Do you know what commands I could use to do that?

**Comment by matt335672** [comment]
Well, I can understand your reticence in some ways, particularly as 20.04 is only just out, and so it's going to be as buggy as it will get. However, `dbus-user-session` is incompatible with XRDP and other remote desktop technologies as I hope the wiki makes clear. You might be able to bodge your way around this by editing the session startup scripts to start a new session dbus for XRDP sessions, but you'll end up with something pretty non-standard. I can't personally recommend this approach, but it's been done in the past with 18.04, with varying degrees of success.

If you have the capability to set up a VM and install another copy of Ubuntu 20.04 to play with, that might give you a bit more confidence that this is the way to go.

With regards XRDP and gdm in 20.04, at present these seem to be incompatible. I'm looking in to that.

I also seem to have hijacked your thread somewhat with the technical detail I've added to this one. I apologise for that - I'll start another thread I think so we can leave this one clear for your original problem.

It's not at all clear to me what will happen with your setup when you terminate RDP sessions, but I suppose the best way to find out is to try it.

I think you could use `loginctl` to list the active sessions, and then try `loginctl terminate-session <ID>` to terminate any RDP sessions. Try that.

**Comment by rdiez** [comment]
I wanted to write down my experiments with Ubuntu MATE 20.04, in case it helps other people. I still did not replace dbus-user-session though.

I tried connecting to a remote desktop with Remmina, and, instead of closing the Remmina window, I logged off properly (with the corresponding menu option). That still did not dispose of the session as it should.

These are the steps to dispose of a lingering desktop:

- Connect with a text console (Ctrl+Alt+F1 or SSH).

- Look at all active sessions:

  loginctl list-sessions

- Identify which session is the user's desktop.
  Note that a user can have other "user services" sessions.

  loginctl user-status

- Terminate the appropriate session:

  This did not work reliably:

    loginctl  terminate-session  <a session ID like c1>

  So I had to resort to:

    loginctl  kill-session  <a session ID like c1>

- Afterwards, a new local or remote desktop session seems to work as expected.

**Comment by matt335672** [comment]
Hi @rdiez 

Do you want any more help with this issue as it stands, or can I close it? Your commentary above will still be available to other people who are searching for it.

Thanks

**Comment by rdiez** [comment]
I think I am done asking questions. Basically, I kind of gave up on remote desktop under Linux. It is too hard for most mortals.

I am not sure about closing this issue though. The problem still exists. I guess that you believe that there is nothing else to be done or to be improved in xrdp to help with this matter, don't you?

**Comment by matt335672** [comment]
I agree there's an issue here. The issue is that currently remote X11 sessions need to have a DBUS session bus which is private to the session. In other words, the sessions need to display different outputs to the command `env |  grep DBUS_SESSION`.

There are a couple of ways to do this with Ubuntu/Debian:-
- the current official recommendation is to replace `dbus-user-session` with `dbus-x11` as outlined in the [wiki](../wiki/Debian-dbus-user-session-package).
- unofficial ways to do this (which you will find in various internet locations) involve starting a session dbus as part of the X session startup scripts.

If you do not do this, you will encounter a significant coupling of sessions, as these will all share a session bus. That's fine for sessions sharing the console and started by GDM, but not for remote sessions.

There may well be further problems once the sessions are decoupled, but these need to be addressed as they are encountered.

I'm very happy to help you get to a working state, but to do that you'll need to follow the official advice. If you do and everything works, that's great. If you do and you've still got problems we can work together on improving the official advice, or the software itself. That's good for all users of xrdp.

Feel free to come back to me on this - the discussions are valuable. I'll close the issue on Weds or thereabouts if I don't hear back from you.

**Comment by rdiez** [comment]
The official advice in the wiki states that there is an unresolved issue with Ubuntu 20.04 LTS. The wiki refers to this bug, which does not really tell any details about this problem. That is the reason why I asked whether xrdp could do something about it.

About the "starting a session dbus as part of the X session startup scripts" that you mentioned, there are some unofficial information on the Internet. Could you document (perhaps in the Wiki) the best / recommended way to achieve this?

**Comment by matt335672** [comment]
Hi @rdiez 
I can see where you're coming from, so thanks for pressing me on this. Yesterday I'd raised a separate fault (#1565) and updated the wiki. You were probably unaware I had.

As I hope the wiki makes clear, the only issues we're aware of are with the GNOME edition of Ubuntu 20.04 LTS. I believe these to be  a consequence of the direction GNOME is moving in, but more low-level investigation is required. 

Rather than just asserting that other editions of 20.04 are unaffected, I decided to test this. I downloaded the MATE version which you are using, installed to on a VM taking defaults, and then did the following:-

- Installed updates and `openssh-server`
- removed `dbus-user-session`
- installed `xrdp` and `xorgxrdp`
- Added the `xrdp` user to the `ssl-cert` group.

I could then log in on both xrdp and the console. Both sessions worked independently of each other, and seemed OK as far as I could tell.

The only oddity I found was that `loginctl` was showing sessions as not cleaning up. I tracked this down to `/usr/libexec/geoclue-2.0/demos/agent` still running. Some more experimentation showed this to be the behaviour whether or not `dbus-user-session` was installed.

So in short, everything seems to work fine with MATE and lightdm provided the information on the wiki is followed. I'll update the wiki with that result.

As for your question about starting a session dbus in the X session startup scripts, this is precisely the behaviour which `dbus-x11` implements if `dbus-user-session` is not installed. The main script which implements this is `/etc/X11/Xsession.d/75dbus_dbus-launch`. This script relies on the `dbus-launch` executable, which again is provided by `dbus-x11`. This package has everything you need. You just need to move `dbus-user-session` out of the way and let it function.

Unless you've got any other evidence to the contrary, I believe (given the above) that the wiki still provides the best advice on how to run XRDP with 20.04 LTS.

Again, feel free to come back to me on any of this.

**Comment by rdiez** [comment]
First of all, thanks again for all the extra information.

> [...]
> Yesterday I'd raised a separate fault (#1565) and updated the wiki. You were probably unaware I had.

I was, thanks for the hint.

The main problem that remains is documentation. Many users will continue to come across error "Could not acquire name on session bus". Just do a quick search on the Internet and you'll find many hits.
If you visit the website at xrdp.org , there is no indication about any documentation, let alone the Wiki. GitHub is for developers, and the first mention I found about the wiki was under the build instructions.

I suggest that you mention the wiki on the official website.

> Added the xrdp user to the ssl-cert group.

The wiki does not mention that this is required, and why. Is this something that Ubuntu should be doing itself?

> The only oddity I found was that loginctl was showing sessions as not cleaning up. 

Mmmm... that could amount to a memory leak per session. But if sessions are always reused, perhaps it does not matter much.

> If you close the RDP client (whatever it is), the session should continue and you should be able to reconnect to it, potentially from another client in another location. This is how the RDP protocol is designed to work and I don't believe there is a way to make the session end automatically if you disconnect.

The behaviour above bothers me. It is not documented as far as I can tell (for example, it is not mentioned in the Wiki). I would welcome an improvement in xrdp so that you can automatically end a session upon disconnection. Or after a few minutes, which assumes that all your reconnection retries have failed. After all, I do not really want a computer to hold user sessions indefinitely (it would use resources that you have not planned for). Or maybe xrdp should distinguish between disconnections (closing the Remmina window) and logging out (from inside the desktop environment). Is this stuff for a new GitHub issue?

**Comment by matt335672** [comment]
I'll agree with you that documentation is somewhat lacking. This is being addressed as and when we can.

Many issues can be the result of how XRDP has been packaged by the distros.
- Adding the `xrdp` user to the `ssl-cert` group is a good example of a packaging issue. The `xrdp` user is a result of Debian packaging (from which Ubuntu is derived). The Debian packaging group have decided that adding the `xrdp` user to the `ssl-cert` group should not be the default - See [Debian issue #860890](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=860890). We have absolutely no control over this here.
- The `loginctl` issue above will lead to a leak, but this is an Ubuntu issue. We are following the PAM rules for creating and destroying sessions and we can't do more than this.

Having some information on the wiki about packaging issues is not ideal for us because we have no control over these issues. Personally I think it's better to have some information here rather than no information at all. 

The behaviour of XRDP regarding sessions is industry standard, and is shared by Microsoft RDP. The behaviour you would like would not be welcome by users who experience regular disconnections over flaky Internet services, or who wish to use the same session from multiple locations (i.e. work and home). If you really want the above behaviour, there is at least one other alternative to XRDP, which is XDMCP in combination with Xvnc. This can be supported by Remmina in VNC client mode (optionally over ssh) and may be more along the lines of what you want.

I hope that's useful - I think these discussions are worthwhile as a source of information for others. Let me know if you've any further comments or questions.

**Comment by metalefty** [comment]
OSS is provided as-is with no warranty. Community driven OSS project is volunteer based. If anyone want something happen, someone has to volunteer and always you also have a chance to be someone.  We always welcome any kind of contributions like issue triage, code, community management, documentation, content, translation, talks, bug reports et cetera.

Each OSS project for example xrdp and Ubuntu are driven separately. As Matt said, we absolutely no control on other projects. I think communication between we xrdp and Ubuntu/Debian is a little bit lacking.  We:re always behind of Ubuntu/Debian and has to follow their change.

I also agree xrdp is not very good regarding documentation however not everything is our fault.

**Comment by rdiez** [comment]
> * Adding the `xrdp` user to the `ssl-cert` group is a good example of a packaging issue. The `xrdp` user is a result of Debian packaging (from which Ubuntu is derived). The Debian packaging group have decided that adding the `xrdp` user to the `ssl-cert` group should not be the default - See [Debian issue #860890](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=860890). We have absolutely no control over this here.

I have dug a little more, and it looks like adding xrdp to that group does not impact "normal" or "simple" usage of xrdp. You only need that if you want to use the SSL certificates, which is not actually mandatory.

I think you should document that on your wiki. You have already typed the information here. After all, you already have a wiki page for system-dependent information:

[Using the console and XRDP together in Debian / Ubuntu / Mint](https://github.com/neutrinolabs/xrdp/wiki/Debian-dbus-user-session-package)

It is normal and helpful for upstream software projects to offer platform-specific information, even if they fall outside of their responsibility. It shows deference for your users. And it actually saves time in the long run, because you do not have to answer the same questions over and over in mailing lists and bug reports.

The alternative is bad publicity for your project (even though perhaps unfounded), as new users immediately encounter problems that many people describe on the Internet for years.

> * The `loginctl` issue above will lead to a leak, but this is an Ubuntu issue. We are following the PAM rules for creating and destroying sessions and we can't do more than this.

I suggest that you document this on the wiki. That makes the issue visible. A mention in a soon-closed bug report like this one is not nearly as effective. I can then create an issue in Ubuntu, or wherever, and link to it. This increases the chance of getting the problem fixed. Visibility is important.

The alternative is users hitting the same issues again and again.


> The behaviour of XRDP regarding sessions is industry standard, and is shared by Microsoft RDP. The behaviour you would like would not be welcome by users who experience regular disconnections over flaky Internet services, or who wish to use the same session from multiple locations (i.e. work and home).
> [...]

Thanks for the infos. However, I find that unconfigurable behaviour unfortunate. And I am not the only one:

"Additionally, xrdp (version 0.9.6) uses the requested screen resolution and color depth to determine if it should reconnect the same session. It used to be configurable (by manually choosing the specific tcp port of the existing session, e.g., 5901) back in v 0.6.something but they decided that was too much power for the users..."

https://serverfault.com/questions/845426/kill-defunct-xrdp-session-xrdp-users-cannot-log-on


> I hope that's useful - I think these discussions are worthwhile as a source of information for others. Let me know if you've any further comments or questions.

I think I am done here. Thanks again for all the support. I have only a personal interest here and I have invested time on this only on a volunteer basis. But I think that's now enough. 8-)

**Comment by matt335672** [comment]
Thanks for your comments and the discussion.
